servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -416,12 +416,19 @@ private Single<C> selectConnection0(final Predicate<C> selector, @Nullable final
                     // Invoke the selector before adding the connection to the pool, otherwise, connection can be
                     // used concurrently and hence a new connection can be rejected by the selector.
                     if (!selector.test(newCnx)) {
-                        // Failure in selection could be temporary, hence add it to the queue and be consistent
-                        // with the fact that select failure does not close a connection.
-                        return newCnx.closeAsync().concat(failed(StacklessConnectionRejectedException.newInstance(
+                        // Failure in selection could be the result of connection factory returning cached connection,
+                        // and not having visibility into max-concurrent-requests, or other threads already selected the
+                        // connection which uses all the max concurrent request count.
+
+                        // If there is caching Propagate the exception and rely upon retry strategy.
+                        Single<C> failedSingle = failed(StacklessConnectionRejectedException.newInstance(
                                 "Newly created connection " + newCnx + " for " + targetResource
                                         + " was rejected by the selection filter.",
-                                RoundRobinLoadBalancer.class, "selectConnection0(...)")));
+                                RoundRobinLoadBalancer.class, "selectConnection0(...)"));
+
+                        // Just in case the connection is not closed add it to the host so we don't lose track,
+                        // duplicates will be filtered out.
+                        return host.addConnection(newCnx) ? failedSingle : newCnx.closeAsync().concat(failedSingle);
```

Comment:
While our `newCnx.closeAsync()` impl always completed, who knows what users applied in their filters. How do you think, in all similar cases inside RRLB, should be just subscribe to` closeAsync()` and then return a failed source instead of concatenating them?